doc/dev/static_type_checking.md

```diff
@@ -0,0 +1,894 @@
+difference.
+
+```python
+from typing import Type
+
+
+class Duck:
+    def quack(self): ...
+
+
+def make_quack(d: Duck) -> None:
+    d.quack()
+
+
+def create_duck(d: Type[Duck]) -> Duck:
+    return d()
+
+
+d = Duck()
+make_quack(d)  # OK
+create_duck(d)  # mypy complains
+
+make_quack(Duck)  # mypy complains
+create_duck(Duck)  # OK
+```
+
+Running mypy shows the expected (and unexpected) output:
+
+```cmd
+main.py:15: error: Argument 1 to "create_duck" has incompatible type "Duck"; expected "Type[Duck]"
+main.py:17: error: Argument 1 to "make_quack" has incompatible type "Type[Duck]"; expected "Duck"
+Found 2 errors in 1 file (checked 1 source file)
+```
+
+### Use typing.TypeAlias when creating a type alias
+
+With [PEP 613](https://peps.python.org/pep-0613/), `typing.TypeAlias` was introduced to make type aliases more obvious
+to the type checker.
+
+```python
+# from typing import TypeAlias Python >=3.10
+from typing_extensions import TypeAlias
+
+CredentialTypes: TypeAlias = Union[AzureKeyCredential, TokenCredential]
+```
+
+This removes ambiguity for the type checker and indicates this isn't a normal variable assignment.
+
+> Note that TypeAlias is backported to older versions of Python by using typing_extensions
+
+### Typing variadic arguments - *args and **kwargs
+
+Many of our client methods take a `**kwargs` and sometimes an `*args` argument.
+
+Consider the below example:
+
+```python
+def begin_operation(*args, **kwargs) -> None:
+    ...
+```
+
+Here, let's assume that `*args` accepts an arbitrary number of positional parameters which must be of type `str`. Let's
+also assume that `**kwargs` accepts operation-specific keyword arguments, including any `azure-core` specific keyword
+arguments. In this case, we can narrow the type for `*args` to `str`. For `**kwargs`, due to the complexity of the type,
+we will leave it as `Any`. Typed this looks like:
+
+```python
+from typing import Any
+
+
+def begin_operation(*args: str, **kwargs: Any) -> None:
+    ...
+```
+
+Note how mypy understands these types:
+
+```python
+from typing import Any
+
+
+def begin_operation(*args: str, **kwargs: Any) -> None:
+    reveal_locals()
+```
+
+```cmd
+main.py:4: note: Revealed local types are:
+main.py:4: note:     args: builtins.tuple[builtins.str]
+main.py:4: note:     kwargs: builtins.dict[builtins.str, Any]]"
+```
+
+`mypy` correctly infers `*args` as a `tuple[str]` and `**kwargs` as a `dict[str, Any]` so we do not need to type these
+as such in the code. If `*args` accepts more than just `str`, a `Union[]` type or `Any` can be used depending on how
+complex the type is.
+
+### Use AnyStr when your parameter accepts both str and bytes
+
+The `typing` module includes a pre-defined `TypeVar` named `AnyStr`. Its definition looks like this:
+
+```python
+from typing import TypeVar
+
+AnyStr = TypeVar('AnyStr', bytes, str)
+```
+
+`AnyStr` can be used in functions which accept or return both bytes and str.
```

Comment:
There is another value to `AnyStr`, it keeps consistent typing with return type

![image](https://user-images.githubusercontent.com/1050156/159375756-4fa32efe-2dc4-42a5-ae38-0cb34d43d143.png)